pkg/controller/container-runtime-config/helpers.go

```diff
@@ -253,87 +253,67 @@ func scopeMatchesRegistry(scope, reg string) bool {
-		root := unionGraph.Find(set.Sources[0]) // == union.Find(set.Sources[x]) for any other x
-		ds, ok := disjointSets[root]
-		if !ok {
-			ds = &[]*apioperatorsv1alpha1.RepositoryDigestMirrors{}
-			disjointSets[root] = ds
+			ds, ok := disjointSets[set.Source]
+			if !ok {
+				ds = &[]*apioperatorsv1alpha1.RepositoryDigestMirrors{}
+				disjointSets[set.Source] = ds
+			}
+			*ds = append(*ds, set)
 		}
-		*ds = append(*ds, set)
 	}
 
-	// Sort the sets of mirrors to ensure deterministic output
-	type dsOrderedKey struct {
-		mapKey  unionNode
-		sortKey string // == The legicographically first mirror in the disjointSets element
-	}
-	dsKeys := []dsOrderedKey{}
-	for mapKey, ds := range disjointSets {
-		sortKey := ""
-		// The build of disjointSets guarantees len(ds) > 0, len(ds[0].Sources) >= 2
-		for _, set := range *ds {
-			for _, src := range set.Sources {
-				if sortKey == "" || sortKey > src {
-					sortKey = src
-				}
-			}
-		}
-		dsKeys = append(dsKeys, dsOrderedKey{
-			mapKey:  mapKey,
-			sortKey: sortKey,
-		})
+	// Sort the sets of mirrors by Source to ensure deterministic output
+	sources := []string{}
+	for key := range disjointSets {
+		sources = append(sources, key)
 	}
-	sort.Slice(dsKeys, func(i, j int) bool {
-		return dsKeys[i].sortKey < dsKeys[j].sortKey
-	})
+	sort.Strings(sources)
 	// Convert the sets of mirrors
-	res := [][]string{}
-	for _, dsKey := range dsKeys {
-		ds := disjointSets[dsKey.mapKey]
+	res := []apioperatorsv1alpha1.RepositoryDigestMirrors{}
+	for _, source := range sources {
+		ds := disjointSets[source]
 		topoGraph := newTopoGraph()
 		for _, set := range *ds {
-			// The build of mirrorSets guarantees len(set.Sources) >= 2, i.e. that this loop runs at least once for every set,
-			// and that every node in topoGraph is implicitly added by topoGraph.AddEdge (there are no unconnected nodes that we would
-			// have to add separately from the edges).
-			for i := 0; i+1 < len(set.Sources); i++ {
-				topoGraph.AddEdge(set.Sources[i], set.Sources[i+1])
+			for i := 0; i+1 < len(set.Mirrors); i++ {
+				topoGraph.AddEdge(set.Mirrors[i], set.Mirrors[i+1])
+			}
+			sourceInGraph := false
+			for _, m := range set.Mirrors {
+				if m == source {
+					sourceInGraph = true
+					break
+				}
 			}
+			if !sourceInGraph {
+				// The build of mirrorSets guarantees len(set.Mirrors) > 0.
+				topoGraph.AddEdge(set.Mirrors[len(set.Mirrors)-1], source)
+			}
+			// Every node in topoGraph, including source, is implicitly added by topoGraph.AddEdge (there are no unconnected nodes that we would
+			// have to add separately from the edges).
 		}
 		sortedRepos, err := topoGraph.Sorted()
 		if err != nil {
 			return nil, err
 		}
-		res = append(res, sortedRepos)
+		if sortedRepos[len(sortedRepos)-1] == source {
+			// We don't need to explicitly include source in the list, it will be automatically tried last per the semantics of sysregistriesv2. Mirrors.
```

Comment:
Is it harmful to leave source in the list? Would it be possible to prevent source from being added to the list rather than removing it later?

Reply:
Right now not removing `source` from `sortedRepos` would result in an unnecessary extra attempt to contact that registry (if the image were not available); it’s not catastrophic but it is unnecessary. Plus, removing `source` from the list results in `registries.conf` more similar to how a human would write it, especially for the trivial case of a single `RepositoryDigestMirrors`.

Yes, not adding the source to the list instead of removing it here would be problematic: Consider
```go
[]RepositoryDigestMirrors{
  {Source: "S", Mirrors: {"A"}} // Expected order of attempts: AS
  {Source: "S", Mirrors: {"BSC"}} // Expected order of attempts: BSC(S) , the second S being redundant/pointless/not applicable
}
```
We need to add the AS edge to `topoGraph` so that the merged order in `sortedRepos` is one of ABSC/BASC; without it, it could end up e.g. BSCA (or maybe BSCA(S)), unnecessarily inconsistent with the input entries.

(Sure, this is all corner cases. Everything about including `Source` as a member of `Mirrors` is a corner case. Originally I had an assert “`Source` is last” in here, then I considered the “`Source` is included in `Mirrors`” counterexample, and it turned out simpler/clearer/more general to just implement that fully than to detect the case and fail.)

Reply:
OK I understand. Thank you for elaborating.